### PR TITLE
[ADP-2956] Upgrade `cardano-node` to 8.1.1 and `ghc`  to `9.2.8`

### PR DESCRIPTION
**Pull request by Unisay**
- [x] bump run time node dependency  to 8.1.1 
- [x] bump compile time cardano node API dependency to ^>=8.2
- [x] remove compile time constraints coming from cardano-node 
- [x] bump ghc to 9.2.8
- [x] bump all tools to latest version
- [x] bump index state of hackage.haskell.org to 2023-06-06T00:00:00Z
- [x] bump index-state of cardano-haskell-packages to 2023-06-05T06:39:32Z
- [x] move stylish-haskell out of tools, to compile old 0.11.0.3 version with old ghc 8.10.7

### Comments

This PR builds on top of 2 other PRs:

- https://github.com/cardano-foundation/cardano-wallet/pull/3959
- https://github.com/cardano-foundation/cardano-wallet/pull/3995

The https://github.com/cardano-foundation/cardano-wallet/pull/3959 is superseded by this one.

This is the **squashed** version of this PR: https://github.com/cardano-foundation/cardano-wallet/pull/3996

This PR introduced a huge regression in the restoration benchmark, which was solved by removing heap profiling. Doing so we have lost the nice memory usage graphs.

### Issue Number

[ADP-2956](https://input-output.atlassian.net/browse/ADP-2956)

### CODE REVIEW

File: .stylish-haskell.yaml [hunk 1]
```diff
@@ -11,7 +11,7 @@ steps:
       empty_list_align: inherit
       list_align: new_line
       list_padding: 4
-      long_list_align: new_line_multiline
+      # long_list_align: new_line_multiline
```
Comment (jonathanknowles) [inline .stylish-haskell.yaml]: Hi @paolino @Unisay!

Is it necessary to comment out this line for the GHC/node bump?

If not, then could we preserve the existing `stylish-haskell` configuration? (i.e., avoid changing the `stylish-haskell` configuration in this PR.)

Justification:

1. Commenting out this setting seems to change the formatting and style of imports for all Haskell files in `cardano-wallet`, not just those related to the node/GHC bump.

2. This causes conflicts with several other in-flight PRs that add/remove imports. If we merge this PR as is, then we'd need to perform additional work to rebase those other PRs and fix them up. (I'd prefer to avoid that, if possible!)

Comment (jonathanknowles) [reply .stylish-haskell.yaml]: UPDATE: after a bit of digging, I found @paolino's bug report on the `stylish-haskell` repo here:

https://github.com/haskell/stylish-haskell/issues/462

Apologies for the noise! 🙏🏻

Comment (paolino) [reply .stylish-haskell.yaml]: Yeah, it's awful, I hope we can move to fourmolu ASAP

File: docker-compose.yml [hunk 2]
```diff
@@ -2,7 +2,7 @@ version: "3.5"
 
 services:
   cardano-node:
-    image: inputoutput/cardano-node:1.35.4
+    image: inputoutput/cardano-node:8.0.0
```
Comment (paolino) [inline docker-compose.yml]: this looks wrong , should it be 8.1.1 ?

File: lib/wallet/exe/mock-token-metadata-server.hs [hunk 3]
```diff
@@ -36,8 +36,28 @@ import "optparse-applicative" Options.Applicative
     , str
     )
 -- See ADP-1910
-import "optparse-applicative" Options.Applicative.Help.Pretty
-    ( hang, indent, line, text, (</>) )
+-- See ADP-1910
```
Comment (paolino) [inline lib/wallet/exe/mock-token-metadata-server.hs]: yeah we should look at ADP-1910 at least `n` times :-)

File: lib/wallet/src/Cardano/Api/Gen.hs [hunk 4]
```diff
@@ -912,8 +920,8 @@ genAddressAnyWithNetworkId genNetworkId' =
     oneof
         [ AddressByron
             <$> genAddressByronWithNetworkId genNetworkId'
-        {- , AddressShelley
-            <$> genAddressShelleyWithNetworkId genNetworkId' -}
+        , AddressShelley
```
Comment (paolino) [inline lib/wallet/src/Cardano/Api/Gen.hs:923]: :+1:

File: lib/wallet/src/Cardano/Api/Gen.hs [hunk 5]
```diff
@@ -1257,7 +1310,10 @@ genTxCertificate =
     oneof
         [ StakeAddressRegistrationCertificate <$> genStakeCredential
         , StakeAddressDeregistrationCertificate <$> genStakeCredential
+        {-
+        For some reason, this constructor is no longer exported:
         , StakeAddressDelegationCertificate <$> genStakeCredential <*> genPoolId
+        -}
```
Comment (paolino) [inline lib/wallet/src/Cardano/Api/Gen.hs]: should we have a ticket to track this ?

File: lib/wallet/src/Cardano/Wallet/Pools.hs [hunk 6]
```diff
@@ -262,8 +262,8 @@ withNodeStakePoolLayer tr settings dbLayer@DBLayer{..} netParams genesisPools ne
 
     -- fixme: needs to be simplified as part of ADP-634
     let NetworkParameters{slottingParameters} = netParams
-        startMetadataThread = forkIOWithUnmask
-            ($ monitorMetadata gcStatus tr slottingParameters dbLayer)
+        startMetadataThread = forkIOWithUnmask $ \f ->
+            f $ monitorMetadata gcStatus tr slottingParameters dbLayer
```
Comment (paolino) [inline lib/wallet/src/Cardano/Wallet/Pools.hs:266]: ghc 9.2 requires this ...

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 7]
```diff
@@ -2537,11 +2508,11 @@ removeDummyInput = \case
                     val
             ShelleyBasedEraConway ->
                 let body' = body
-                        { Conway.inputs =
+                        {- Conway.inputs =
```
Comment (paolino) [inline lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]: this should have a HF blocking ticket associated ?

File: lib/wallet/test/unit/Cardano/Wallet/Primitive/SlottingSpec.hs [hunk 8]
```diff
@@ -65,6 +65,7 @@ import Data.Generics.Internal.VL.Lens
     ( (^.) )
 import Data.Quantity
     ( Quantity (..) )
+import Data.SOP.Counting
```
Comment (paolino) [inline lib/wallet/test/unit/Cardano/Wallet/Primitive/SlottingSpec.hs:68]: we should restrict to `exactlyTwo`

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 9]
```diff
@@ -519,36 +532,37 @@ spec = do
     describe "Sign transaction" $ do
         -- TODO [ADP-2849] The implementation must be restricted to work only in
         -- 'RecentEra's, not just the tests.
-        spec_forAllRecentEras
+        spec_forAllRecentErasPedningConway
```
Comment (paolino) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: Pedning -> Pending

File: scripts/make_release.sh [hunk 10]
```diff
@@ -23,7 +23,7 @@ SCRIPT=$(realpath "$0")
 # Release tags must follow format vYYYY-MM-DD.
 GIT_TAG="v2023-04-14"
 OLD_GIT_TAG="v2022-12-14"
-CARDANO_NODE_TAG="1.35.4"
+CARDANO_NODE_TAG="8.0.0"
```
Comment (paolino) [inline scripts/make_release.sh]: should this be 8.1.1 ?

File: .stylish-haskell.yaml (hunk 1 above)
Comment (paolino) [reply .stylish-haskell.yaml]: We manage to roll back the change and still have stylish-haskell 0.11.0.3 in scope

File: lib/wallet/src/Cardano/Api/Gen.hs (hunk 4 above)
Comment (paolino) [reply lib/wallet/src/Cardano/Api/Gen.hs:923]: solved

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs (hunk 7 above)
Comment (paolino) [reply lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]: solved

File: lib/wallet/test/unit/Cardano/Wallet/Primitive/SlottingSpec.hs (hunk 8 above)
Comment (paolino) [reply lib/wallet/test/unit/Cardano/Wallet/Primitive/SlottingSpec.hs:68]: solved

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs (hunk 9 above)
Comment (paolino) [reply lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: solved

File: scripts/make_release.sh (hunk 10 above)
Comment (paolino) [reply scripts/make_release.sh]: solved

File: lib/primitive/test/spec/Cardano/Wallet/Primitive/CollateralSpec.hs [hunk 11]
```diff
@@ -698,8 +693,8 @@ spec = do
         describe "generators and shrinkers" $ do
             it "generates values with sufficient coverage" $
                 property prop_genAddress_coverage
-            it "shrink maintains validity" $
-                property prop_simplifyAddress_validAddress
+            -- it "shrink maintains validity" $
+            --     property prop_simplifyAddress_validAddress
```
Comment (Anviking) [inline lib/primitive/test/spec/Cardano/Wallet/Primitive/CollateralSpec.hs]: I managed to easily re-enable 
- `prop_simplifyAddress_validAddress`
- `prop_addressType_stake`

Will push. The rest didn't seem as straight-forward, but I think it is fine to leave.

File: lib/wallet/api/http/Cardano/Wallet/Launch/Cluster.hs [hunk 12]
```diff
@@ -907,41 +925,34 @@ generateGenesis dir systemStart initialFunds addPoolsToGenesis = do
...
-            , _maxBBSize = 239_857
-            , _maxBHSize = 217_569
-            , _maxTxSize = 16_384
-
-            , _minPoolCost = Ledger.Coin 0
-
-            , _extraEntropy = Ledger.NeutralNonce
-
+    let pparams =
+            Ledger.emptyPParams
```
Comment (Anviking) [inline lib/wallet/api/http/Cardano/Wallet/Launch/Cluster.hs:929]: We now won't be notified if a new field is added which we need to provide a more sensible value for what `emptyPParams` does. Looking at `emptyPParams` it does seem some values are more sensible than just "0" though, so hopefully this isn't a problem we're going to run into in the future.

https://github.com/input-output-hk/cardano-ledger/blob/2a03a3ee0b8b401552e985b7035c3dc382ff2f29/eras/babbage/impl/src/Cardano/Ledger/Babbage/PParams.hs#L317-L335

File: lib/wallet/src/Cardano/Wallet/Read/Primitive/Coin.hs [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+import GHC.Stack
+    ( HasCallStack )
+
+import qualified Cardano.Ledger.Coin as Ledger
+import qualified Cardano.Wallet.Primitive.Types.Coin as Coin
+import qualified Cardano.Wallet.Primitive.Types.Coin as Wallet
+
+unsafeFromLedger :: HasCallStack => Ledger.Coin -> Wallet.Coin
+unsafeFromLedger (Ledger.Coin c) = Coin.unsafeFromIntegral c
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Read/Primitive/Coin.hs]: This already exists: https://github.com/cardano-foundation/cardano-wallet/blob/master/lib/wallet/src/Cardano/Wallet/Shelley/Compatibility/Ledger.hs#L145-L153

Comment (Unisay) [reply lib/wallet/src/Cardano/Wallet/Read/Primitive/Coin.hs]: 👌🏼 I'll fix it.

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs [hunk 14]
```diff
@@ -197,15 +196,19 @@ import Text.Pretty.Simple
 import qualified Cardano.Api as Cardano
 import qualified Cardano.Api.Byron as Cardano
 import qualified Cardano.Api.Shelley as Cardano
+import Cardano.Ledger.Alonzo.Core
+    ( ppCollateralPercentageL, ppMaxCollateralInputsL )
+import Cardano.Ledger.Api
+    ( outputsTxBodyL, ppMaxTxSizeL, ppMaxValSizeL )
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs]: Wrong section

File: lib/wallet/test-common/Cardano/Wallet/DummyTarget/Primitive/Types.hs [hunk 15]
```diff
@@ -274,15 +276,15 @@ babbageMainnetProtocolParameters = C.ProtocolParameters
...
+    , C.protocolParamUTxOCostPerWord = Just $ C.Lovelace 34_482
+    , C.protocolParamUTxOCostPerByte = Just $ C.Lovelace 4_310
     , C.protocolParamCostModels =
         mempty
         -- TODO: Include a Plutus cost model here.
     , C.protocolParamPrices =
         Just $ C.ExecutionUnitPrices
             { C.priceExecutionSteps = 7.21e-5
-            , C.priceExecutionMemory = 0.0577
+            , C.priceExecutionMemory = 0.057_7
```
Comment (Anviking) [inline lib/wallet/test-common/Cardano/Wallet/DummyTarget/Primitive/Types.hs:287]: TBH, not sure I find this more readable

File: nix/haskell.nix [hunk 16]
```diff
@@ -54,9 +54,12 @@ CHaP: haskell-nix: nixpkgs-recent: nodePkgs: haskell-nix.cabalProject' [
...
       # cross compilation (e.g. to windows)
       setGitRevPostInstall = setGitRevPostInstall' config.gitrev;
-      setGitRevPostInstall' = gitrev: ''
-        ${pkgs.buildPackages.haskellBuildUtils}/bin/set-git-rev "${gitrev}" $out/bin/*
-      '';
+      setGitRevPostInstall' = gitrev: '' '';
+      # The following is commented out because it causes error with
+      # 'packages.cardano-node.package.identifier.name' not being defined.
+      # setGitRevPostInstall' = gitrev: ''
+      #   ${pkgs.buildPackages.haskellBuildUtils}/bin/set-git-rev "${gitrev}" $out/bin/*
```
Comment (Anviking) [inline nix/haskell.nix:61]: 🆗

File: nix/haskell.nix [hunk 17]
```diff
@@ -92,9 +95,10 @@ CHaP: haskell-nix: nixpkgs-recent: nodePkgs: haskell-nix.cabalProject' [
         filter = lib.cleanSourceFilter;
       };
 
+      indexState = "2023-06-04T22:30:25Z";
     in {
       name = "cardano-wallet";
-      compiler-nix-name = "ghc8107";
+      compiler-nix-name = "ghc928";
```
Comment (Anviking) [inline nix/haskell.nix:101]: 🎉

**Review by Anviking — APPROVED** [review]
Nice work! Lgtm modulo CI passing

**Review by paolino — APPROVED** [review]

File: lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Features/Fee.hs [hunk 18]
```diff
@@ -15,9 +13,8 @@ import Cardano.Wallet.Read.Eras
 import Cardano.Wallet.Read.Tx.Fee
     ( Fee (..), FeeType )
 
-import qualified Cardano.Ledger.Coin as SL
 import qualified Cardano.Wallet.Primitive.Types.Coin as W
-import qualified Cardano.Wallet.Primitive.Types.Coin as Coin
+import qualified Cardano.Wallet.Shelley.Compatibility.Ledger as Ledger
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Features/Fee.hs:17]: The plan was to inline `*.Compatibility` into these more specialized modules — but this is still good enough for merging.

File: lib/wallet/test/unit/Cardano/Wallet/Address/Discovery/RandomSpec.hs [hunk 19]
```diff
@@ -8,6 +8,7 @@
 {-# LANGUAGE TypeApplications #-}
 
 {-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
```
Comment (HeinrichApfelmus) [inline lib/wallet/test/unit/Cardano/Wallet/Address/Discovery/RandomSpec.hs:11]: I think that this warning is somewhat out of scope for this pull request. 😅  But it's done now.

**Review by HeinrichApfelmus — APPROVED** [review]
Looks good to me — I can find no obvious defects! 😊

File: .stylish-haskell.yaml (hunk 1 above)
Comment (jonathanknowles) [reply .stylish-haskell.yaml]: > We manage to roll back the change and still have stylish-haskell 0.11.0.3 in scope

Wow, that's great news! 🍾

Thanks! ❤️

File: lib/wallet/test/unit/Cardano/Wallet/Address/Discovery/RandomSpec.hs (hunk 19 above)
Comment (paolino) [reply lib/wallet/test/unit/Cardano/Wallet/Address/Discovery/RandomSpec.hs:11]: it's for the specs to use partial functions ?

**Review by HeinrichApfelmus — APPROVED** [review]
Looks like a fix to the serialization issue to me. 😊

File: lib/wallet/test/unit/Cardano/Wallet/Address/Discovery/RandomSpec.hs (hunk 19 above)
Comment (HeinrichApfelmus) [reply lib/wallet/test/unit/Cardano/Wallet/Address/Discovery/RandomSpec.hs:11]: Yes. (I'm not fond of that warning.) What I meant was that the specs worked before without this `GHC_OPTIONS` pragma, but now they seem to need it, and this particular change seems out of scope to me as it touches a lot of modules.

But they have been touched now, I'm fine with proceeding.
